src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -0,0 +1,823 @@
+                               err = readClient->SendRequest(readParams);
+                           } else {
+                               // SendAutoResubscribeRequest cleans up the params, even on failure.
+                               attributePath.release();
+                               eventPath.release();
+                               err = readClient->SendAutoResubscribeRequest(std::move(readParams));
+                           }
+
+                           if (err != CHIP_NO_ERROR) {
+                               dispatch_async(self.queue, ^{
+                                   [self _handleSubscriptionError:[MTRError errorForCHIPErrorCode:err]];
+                               });
+
+                               return;
+                           }
+
+                           // Callback and ReadClient will be deleted when OnDone is called or an error is
+                           // encountered.
+                           callback->AdoptReadClient(std::move(readClient));
+                           callback.release();
+                       }];
+}
+
+// TODO: support more than one subscriber / subscription delegate
+//   * reconcile second+ subscriptions with different min and max intervals
+//     (save min/max intervals for each subscriber and resubscribe when they change)
+//   * events should be cached and replayed for second+ subscribers
+- (void)subscribeWithDelegate:(id<MTRDeviceSubscriptionDelegate>)delegate
+                        queue:(dispatch_queue_t)queue
+                  minInterval:(uint16_t)minInterval
+                  maxInterval:(uint16_t)maxInterval
+                       params:(MTRSubscribeParams *)params
+{
+    // Save a weak reference to delegate, so if/when the object goes away, they are automatically unsubscribed
+    MTRWeakReference<id<MTRDeviceSubscriptionDelegate>> * weakDelegate = [MTRWeakReference weakReferenceWithObject:delegate];
+    [_subscribers addObject:weakDelegate];
+
+    // perform the actual subscription
+    [self subscribeWithMinInterval:minInterval maxInterval:maxInterval params:params];
+}
+
+#pragma mark Device Interactions
+- (NSDictionary<NSString *, id> *)readAttributeWithEndpointId:(NSNumber * _Nullable)endpointId
+                                                    clusterId:(NSNumber * _Nullable)clusterId
+                                                  attributeId:(NSNumber * _Nullable)attributeId
+                                                       params:(MTRReadParams * _Nullable)params
+{
+    // Create work item, set ready handler to perform task, then enqueue the work
+    MTRAsyncCallbackQueueWorkItem * workItem = [[MTRAsyncCallbackQueueWorkItem alloc] initWithQueue:_queue];
+    MTRAsyncCallbackReadyHandler readyHandler = ^(MTRDevice * device, NSUInteger retryCount) {
+        [self connectAndPerformAsync:^(MTRBaseDevice * baseDevice) {
```

Comment:
Why do you need the `connectAndPerformAsync`, given that `baseDevice readAttributeWithEndpointId` will do the connecting itself?